plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -3984,7 +3984,12 @@ private static Optional<CatalogSchemaTableName> redirectTableToHudi(Optional<Str
     @Override
     public boolean allowSplittingReadIntoMultipleSubQueries(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        SchemaTableName tableName = ((HiveTableHandle) tableHandle).getSchemaTableName();
+        // dont split to subqueries if tableHandle is systemTableHandle
+        if (!(tableHandle instanceof HiveTableHandle hiveTableHandle)) {
```

Comment:
I would check for the `SystemTableHandle` in the `io.trino.metadata.MetadataManager#allowSplittingReadIntoMultipleSubQueries`. This avoids copying this check to multiple places.

Reply:
You can't pass any table handle to any connector. Every connector expects only its own `TableHandle` class. The casts like that are all over the `HiveMetadata`.  The `CatalogHandle` actually has `io.trino.spi.connector.CatalogHandle#type` that in this case is set to `io.trino.spi.connector.CatalogHandle.CatalogHandleType#SYSTEM`.